src/track/beats.cpp

```diff
@@ -661,6 +715,210 @@ std::optional<BeatsPointer> Beats::tryScale(BpmScale scale) const {
             m_subVersion));
 }
 
+std::optional<BeatsPointer> Beats::tryAdjustTempo(
+        audio::FramePos position, TempoAdjustment adjustment) const {
+    auto markers = m_markers;
+
+    // Retrieve the Beat Marker before the current position.
+    //
+    // Hence, we first get the marker *after* the position, then decrement the
+    // iterator.
+    // If there is no marker before the current position, the marker after the
+    // position can be used because the beats before the first marker are
+    // interpolated using the same beat length as the beats after the first
+    // marker.
+    auto markerIt = std::upper_bound(markers.begin(),
+            markers.end(),
+            position,
+            isFramePosLessThanBeatMarker);
+    if (markerIt != markers.begin() && position < m_lastMarkerPosition) {
+        markerIt--;
+    }
+
+    // At this point, `markerIt` points to the beat marker that we need to
+    // modify (or `cend()`, which means that we need to modify the BPM of the
+    // last marker).
+    Bpm lastMarkerBpm = m_lastMarkerBpm;
+    if (markerIt == markers.end()) {
+        lastMarkerBpm += (adjustment == TempoAdjustment::Faster) ? kBpmAdjustStep : -kBpmAdjustStep;
+        if (!lastMarkerBpm.isValid()) {
+            qWarning() << "Beats: Tempo adjustment would result in invalid Bpm!";
+            return std::nullopt;
+        }
+    } else {
+        const auto marker = *markerIt;
+        const int adjustedBeatsTillNextMarker = marker.beatsTillNextMarker() +
+                ((adjustment == TempoAdjustment::Faster)
+                                ? marker.beatsPerBar()
+                                : -marker.beatsPerBar());
+
+        if (adjustedBeatsTillNextMarker < marker.beatsPerBar()) {
+            qWarning() << "Beats: Tempo adjustment would result in a marker "
+                          "with less than the minimum beats in one bar!";
+            return std::nullopt;
+        }
+
+        markerIt = markers.erase(markerIt);
+        markerIt = markers.emplace(markerIt,
```

Comment:
Reusing the name "markerIt" is confusing.

Reply:
Should I really declare a new variable each time I advance an iterator?

`erase` removes the element at the iterator position and returns an iterator pointing to the element behind it. `emplace` inserts an new element in front of the iterator position and returns an iterator pointing to the new element.

So what this does it it replaces the element at the iterator position with a new element.

Reply:
The issue here is that it is not instantly clear how the iterator is advanced. But this naming issue is a minor nit pick anyway.

Would it be possible to edit the iterator in place? 

How is verified that the beats are always in consecutive order? Can this her an issue?

Reply:
> The issue here is that it is not instantly clear how the iterator is advanced.

It requires to know how the `std::vector` API works. `std::vector` is one of the most containers in the C++ standard library. I don't see an issue with that.

> Would it be possible to edit the iterator in place?

No.